### PR TITLE
Core: Update version-hint.txt atomically

### PR DESCRIPTION
**Pull request by lcspinter**
Under issue #1496 it was already discussed that we should find a way to update version-hint.txt atomically. At the moment, there is no know operation that would support atomic updates. Based on this I think our best solution would be:
1. Create a new file
2. Delete old version-hint.txt
3. Move the new file to version-hint.txt

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java [hunk 1]
```diff
@@ -294,13 +294,25 @@ private void writeVersionHint(int versionToWrite) {
     Path versionHintFile = versionHintFile();
     FileSystem fs = getFileSystem(versionHintFile, conf);
 
-    try (FSDataOutputStream out = fs.create(versionHintFile, true /* overwrite */)) {
-      out.write(String.valueOf(versionToWrite).getBytes(StandardCharsets.UTF_8));
+    try {
+      if (fs.exists(versionHintFile)) {
+        Path tempVersionHintFile = metadataPath("version-hint.temp" );
```
Comment (pvary) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:299]: I would go for an uniq name here so 2 concurrent commits will not interfere

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java [hunk 2]
```diff
@@ -294,13 +294,25 @@ private void writeVersionHint(int versionToWrite) {
...
+        fs.rename(tempVersionHintFile, versionHintFile);
+      } else {
+        writeFileToPath(fs, versionHintFile);
+      }
     } catch (IOException e) {
       LOG.warn("Failed to update version hint", e);
     }
   }
 
+  private void writeFileToPath(FileSystem fs, Path path) throws IOException{
```
Comment (pvary) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:311]: Formatting checks might fail here.

I have learned that it is worth to run checkstyle checks before committing:
```
./gradlew build -x test
```

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java [hunk 3]
```diff
@@ -294,13 +294,25 @@ private void writeVersionHint(int versionToWrite) {
...
+    FSDataOutputStream out = fs.create(path, true /* overwrite */);
+    out.write(String.valueOf(path).getBytes(StandardCharsets.UTF_8));
```
Comment (pvary) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:313]: We might want to close the stream here (try with resources block can help)

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java [hunk 4]
```diff
@@ -294,13 +294,25 @@ private void writeVersionHint(int versionToWrite) {
...
-    try (FSDataOutputStream out = fs.create(versionHintFile, true /* overwrite */)) {
-      out.write(String.valueOf(versionToWrite).getBytes(StandardCharsets.UTF_8));
+    try {
+      if (fs.exists(versionHintFile)) {
+        Path tempVersionHintFile = metadataPath("version-hint.temp" );
+        writeFileToPath(fs, tempVersionHintFile);
+        fs.delete(versionHintFile, false);
+        fs.rename(tempVersionHintFile, versionHintFile);
+      } else {
+        writeFileToPath(fs, versionHintFile);
```
Comment (pvary) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:304]: Again concurrency can be problematic here

**Comment by HeartSaVioR** [comment]
If I understand correctly, these operations should be applied regardless of existing version file - consider the case when two concurrent writers both see there's no existing version file and try to write to the path directly.

That said, delete -> rename can be (should be?) replaced with rename with overwrite = true. If I'm not missing anything, this ensures last one wins and partial file is not exposed.

**Comment by pvary** [comment]
> That said, delete -> rename can be (should be?) replaced with rename with overwrite = true. If I'm not missing anything, this ensures last one wins and partial file is not exposed.

I think we could not use rename with overwrite since it behaves differently on different [FileSystems](https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/filesystem/filesystem.html#boolean_rename.28Path_src.2C_Path_d.29):
```
Destination exists and is a file

Renaming a file atop an existing file is specified as failing, raising an exception.
- Local FileSystem : the rename succeeds; the destination file is replaced by the source file.
- HDFS : The rename fails, no exception is raised. Instead the method call simply returns false.
```

That is why we decided to use delete + rename which works consistently for every FS

**Comment by lcspinter** [comment]
@pvary Thanks for the review. I've committed an improved version of the fix.

**Comment by HeartSaVioR** [comment]
> - HDFS : The rename fails, no exception is raised. Instead the method call simply returns false.

I guess it's not explaining the case when overwrite = true. (HDFS is breaking contract then.)

delete -> rename would work if we simply do delete regardless of when the file exists or not, so not a big deal though. Just wanted to try to ensure the last one wins, but on concurrent writers it wouldn't be weird anyone wins.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java [hunk 5]
```diff
@@ -295,22 +295,19 @@ private void writeVersionHint(int versionToWrite) {
...
     } catch (IOException e) {
       LOG.warn("Failed to update version hint", e);
     }
   }
 
-  private void writeFileToPath(FileSystem fs, Path path) throws IOException{
-    FSDataOutputStream out = fs.create(path, true /* overwrite */);
-    out.write(String.valueOf(path).getBytes(StandardCharsets.UTF_8));
+  private void writeVersionToPath(FileSystem fs, Path path, int versionToWrite) throws IOException {
+    try (FSDataOutputStream out = fs.create(path, true /* overwrite */)) {
```
Comment (HeartSaVioR) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:308]: I wouldn't set overwrite to true to ensure this doesn't break anything. (In practice it won't, but it should be much clearer with overwrite = false.)

Comment (lcspinter) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java]: Thanks for the review. Done

**Comment by pvary** [comment]
+1

**Review by HeartSaVioR — APPROVED** [review]
LGTM

**Comment by rdblue** [comment]
If there is a rename with overwrite option, then I would prefer to use that for the cases where it _is_ atomic. That would be safer. (Side note: local FS doesn't provide atomic rename, even without overwrite, so you shouldn't be using Hadoop tables with it except for tests.)

Also, we don't care whether the version hint is completely up to date. We just want to be close to the real version. Being a version or two behind is okay.

**Comment by lcspinter** [comment]
Since the FileSystem API doesn't provide atomic renames, that's why we choose to create new file + delete old file + rename new file path. 
@rdblue If you don't have any more remarks, could you please push this change to master? Thanks

**Comment by lcspinter** [comment]
@rdblue When you have some time, could you please review my change? Thanks

**Comment by rdblue** [comment]
Does the file system API provide a rename with overwrite option? If so, we should use that regardless of whether it is atomic. If not, then we should move forward with delete and then rename.

**Comment by lcspinter** [comment]
@rdblue I agree, rename with overwrite would be the best solution but on HDFS the rename fails without raising any exception if destination exists.

**Comment by lcspinter** [comment]
@rdblue According to the Filesystem rename API doc
Local FileSystem : the rename succeeds; the destination file is replaced by the source file.
HDFS : The rename fails, no exception is raised. Instead the method call simply returns false.

What if I try the rename with overwrite, and if it returns false I fallback to delete and rename? What do you think?

**Comment by HeartSaVioR** [comment]
To clarify, the API I mentioned was `FileContext.rename(src, dst, options)`, not `FileSystem.rename(src, dst)`, which the filesystem doc documents.

I don't have HDFS cluster now, but the result of the rename operation against local filesystem via FileContext is quite different from the document says.

Below is the code you can run with spark-shell against Hadoop 2.7 & Hadoop 3.2.

```

import org.apache.hadoop.fs.{FileContext, Path}
import org.apache.hadoop.fs.Options.Rename
import org.apache.hadoop.fs.permission.FsPermission

val context = FileContext.getFileContext()

// assuming you have files `unit-tests.log` and `unit-tests-succeed.log` (different file size) in /tmp

val setupPath = new Path("/tmp/unit-tests-succeed.log")
val anotherFilePath = new Path("/tmp/unit-tests.log")
val sourceDirPath = new Path("/tmp/rename-experiment-src")
val destDirPath = new Path("/tmp/rename-experiment-dst")
val anotherFileSourcePath = new Path("/tmp/rename-experiment-src/unit-tests.log")
val sourcePath = new Path("/tmp/rename-experiment-src/unit-tests-succeed.log")
val destPath = new Path("/tmp/rename-experiment-dst/unit-tests-succeed.log")

// remove directories
context.delete(sourceDirPath, true)
context.delete(destDirPath, true)
context.mkdir(sourceDirPath, FsPermission.getDirDefault(), true)
context.mkdir(destDirPath, FsPermission.getDirDefault(), true)

// setup file
context.util.copy(setupPath, sourcePath)

// the file got moved
context.rename(sourcePath, destPath)

// check whether the file is moved
println(s"src path: ${context.util.exists(sourcePath)}")
println(s"dest path: ${context.util.exists(destPath)}")
println(s"content summary on dest path: ${context.util.getContentSummary(destPath)}")

// re-setup file
context.util.copy(setupPath, sourcePath)

// re-rename -> this will throw exception as file already exists
context.rename(sourcePath, destPath)

// setup another file
context.util.copy(anotherFilePath, anotherFileSourcePath)

// re-rename with overwrite option -> this will not throw exception
context.rename(anotherFileSourcePath, destPath, Rename.OVERWRITE)

// check whether the file is moved
println(s"src path: ${context.util.exists(anotherFileSourcePath)}")
println(s"dest path: ${context.util.exists(destPath)}")
println(s"content summary on dest path: ${context.util.getContentSummary(destPath)}")
```

It correctly fails on existing file in destination, and correctly overwrites the new file if the overwrite option is provided.

I also looked through the code path on how namenode handles rename, and it redirects me to `DistributedFileSystem.rename` which javadoc says it guarantees atomicity.

> rel/release-2.7.4

https://github.com/apache/hadoop/blob/cd915e1e8d9d0131462a0b7301586c175728a282/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DistributedFileSystem.java#L647-L653

> rel/release-3.2.0

https://github.com/apache/hadoop/blob/e97acb3bd8f3befd27418996fa5d4b50bf2e17bf/hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DistributedFileSystem.java#L892-L898

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java [hunk 6]
```diff
@@ -294,13 +294,22 @@ private void writeVersionHint(int versionToWrite) {
...
+      writeVersionToPath(fs, tempVersionHintFile, versionToWrite);
+      fs.delete(versionHintFile, false);
+      fs.rename(tempVersionHintFile, versionHintFile);
     } catch (IOException e) {
       LOG.warn("Failed to update version hint", e);
     }
   }
 
+  private void writeVersionToPath(FileSystem fs, Path path, int versionToWrite) throws IOException {
+    try (FSDataOutputStream out = fs.create(path, false)) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:308]: Boolean parameters need inline comments to explain what they are. What does `false` here do?

**Comment by rdblue** [comment]
Good to understand the points about rename with overwrite flag. Since there isn't an overwrite flag in the API we use, let's go with delete and rename for this. The current approach looks good to me.

**Comment by lcspinter** [comment]
@rdblue Thanks for the review. I updated the code with an inline comment.

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java [hunk 7]
```diff
@@ -294,13 +294,22 @@ private void writeVersionHint(int versionToWrite) {
     Path versionHintFile = versionHintFile();
     FileSystem fs = getFileSystem(versionHintFile, conf);
 
-    try (FSDataOutputStream out = fs.create(versionHintFile, true /* overwrite */)) {
-      out.write(String.valueOf(versionToWrite).getBytes(StandardCharsets.UTF_8));
+    try {
+      Path tempVersionHintFile = metadataPath(UUID.randomUUID().toString() + "-version-hint.temp");
+      writeVersionToPath(fs, tempVersionHintFile, versionToWrite);
+      fs.delete(versionHintFile, false);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:300]: This one needs a comment to clarify `false` as well.

Comment (lcspinter) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java]: Thanks. Fixed it.

**Comment by rdblue** [comment]
Thanks, @lcspinter! Merged.

**Comment by lcspinter** [comment]
@rdblue Thanks!

resolves #1496

**Comment by sylvon** [comment]
@lcspinter @rdblue we observed a few times in our job that when the program exits say due to OOM right after the delete but before the rename, we will no longer be able to find the table. What we have to do is manually find the latest version and craft a version-hint file by hand. Since we do this delete and rename operation, can we also generate a version-hint by file listing when the file is not found?

**Pull request merged**